smolyay/basis/basis.py

```diff
@@ -0,0 +1,249 @@
+
+
+class BasisFunctionSet(abc.ABC):
+    """Set of basis functions
+    Creates a framework for a set of basis functions to be used as the
+    building blocks for a surrogate model.
+    ``basis_set`` is a list of BasisFunction objects
+    ``sample_flag`` is a list of boolean equal in length to basis_set and
+    an index in the list is true if the object in basis_set at the same
+    index is one that points should be sampled from
+    ``basis_function`` is the BasisFunction child class that populates
+    basis_set
+    ''all_points'' 1D points taken from the BasisFunction objects specified
+    by sample_flag
+    ``need_update`` is a boolean parameter that specifies when properties
+    of the class need to be recomputed
+
+    """
+
+    def __init__(self,sample_flag,basis_function):
+        self._sample_flag = sample_flag
+        self._basis_function = basis_function
+        self._basis_set = []
+        self._all_points = []
+        self._need_update = True
+
+    @property
+    def all_points(self):
+        """All points for basis function set at some level of precision"""
+        if self._need_update:
+            self._update()
+        return self._all_points
+
+    @property
+    def basis_set(self):
+        """list of BasisFunction objects"""
+        if self._need_update:
+            self._update()
+        return self._basis_set
+
+    @property
+    def sample_flag(self):
+        """list of what objects in basis_set to sample points from"""
+        return self._sample_flag
+
+    @sample_flag.setter
+    def sample_flag(self,sample_flag):
+        self._sample_flag = sample_flag
+        self._need_update = True
+
+    @property
+    def basis_function(self):
+        """BasisFunction class used in basis_set"""
+        return basis_function
+
+    @basis_function.setter
+    def basis_function(self,basis_function):
+        self._basis_function = basis_function
+        self._need_update = True
+
+    @abc.abstractmethod
+    def _update(self):
+        """Updates all_points based on parameters"""
+        pass
+
+
+class RecurrenceSet(BasisFunctionSet):
+    """Set of basis functions related via a recurrence relation
+
+    The set of basis functions this class uses are assumed to be part
+    of a sequential series of functions, where the index of a function
+    in basis_set is the only required input of the basis_function. Such
+    functions can be described as a recurrence relation where the nth
+    function of the sequence is a combination of previous terms.
+    """
+
+    def __init__(self,sample_flag,basis_function):
+        """Initialization of parameters"""
+        super().__init__(sample_flag,basis_function)
+        self._update()
+
+    def _update(self):
+        """Update basis_set and get points"""
+        self._basis_set = []
+        self._all_points = []
+
+        for i in range(0,len(self._sample_flag)):
+            self._basis_set.append(self._basis_function(i))
```

Comment:
I also don't quite follow this, couldn't this be passed to the constructor as a list of concrete objects, rather than the class to be initialized?

Reply:
Technically, yes. basis_set could be an input for the constructor. However, given that there was only one child of BasisFunction, I didn't see the need to make creating basis_set something that needed to be calculated beforehand.

I can have BasisFunctionSet require basis_set as an input instead if that would be better.

Reply:
One advantage of doing that would be that `sample_indexes` (or whatever it is called) would refer to the index of the basis function in this list, rather than a constructor argument.

Reply:
Actually: this is going to be important. The general `BasisFunction` does not take `n` as an input, so this line could actually fail for other potential children.

Reply:
Well, the fact that it assumed that n was an input needed for `basis_function` is why it's called RecurrenceSet. It assumed that `basis_function` can be described by a recurrence relation and thus the objects need to know where they are in that recurrence relation.

Reply:
You would need to have a new type of `BasisFunction` to enforce this type check, but I think it is probably not a good idea. You don't want to force a class to have a specific constructor, and this would mean the class can only have one argument (`n`) in its constructor.